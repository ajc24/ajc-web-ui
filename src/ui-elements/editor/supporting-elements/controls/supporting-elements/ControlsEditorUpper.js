import React from 'react';
import PropTypes from 'prop-types';
import HorizontalButtonList from '../../../../form/HorizontalButtonList';
import AddNewItemDialog from '../../dialogs/AddNewItemDialog';
import LoadingNewItemDialog from '../../dialogs/LoadingNewItemDialog';
import '../../../../../css/common.css';

/**
 * The ControlsEditorUpper component renders the controls / buttons for the upper section of the editor panel.
 */
class ControlsEditorUpper extends React.Component {
  constructor(props) {
		super(props);
    this.state = {
      dialogAddNewItem: false,
      dialogLoadingNewItem: false,
    };
    this.handleClickConfirmAddItem = this.handleClickConfirmAddItem.bind(this);
    this.hideAddNewItemDialog = this.hideAddNewItemDialog.bind(this);
    this.hideLoadingNewItemDialog = this.hideLoadingNewItemDialog.bind(this);
    this.showAddNewItemDialog = this.showAddNewItemDialog.bind(this);
    this.showLoadingNewItemDialog = this.showLoadingNewItemDialog.bind(this);
  }

  /**
   * Handles click actions to confirm adding a new item from the dialog dropdown selection
   */
  handleClickConfirmAddItem() {
    /* Set the user value to the hidden input field */
    document.querySelector('span[id="add-new-item-user-value"]').textContent = document.querySelector('select[id="add-new-item-dropdown-menu-id"]').value;
    /* Invoke the functionality to add the new editor item */
    this.showLoadingNewItemDialog();
    setTimeout(() => {
      this.props.addEditorItem();
      this.hideLoadingNewItemDialog();
    }, this.props.reRenderAllowance);
  }

  /**
   * Marks the add new item dialog as hidden.
   */
  hideAddNewItemDialog() {
    const newState = Object.assign({}, this.state, {
      dialogAddNewItem: false,
    });
    this.setState(newState);
  }

  /**
   * Marks the loading dialog for adding a new item as hidden.
   */
   hideLoadingNewItemDialog() {
    const newState = Object.assign({}, this.state, {
      dialogLoadingNewItem: false,
    });
    this.setState(newState);
  }

  /**
   * Marks the add new item dialog as visible.
   */
   showAddNewItemDialog() {
    const newState = Object.assign({}, this.state, {
      dialogAddNewItem: true,
    });
    this.setState(newState);
  }

  /**
   * Marks the loading dialog for adding a new item as visible.
   */
   showLoadingNewItemDialog() {
    const newState = Object.assign({}, this.state, {
      dialogAddNewItem: false,
      dialogLoadingNewItem: true,
    });
    this.setState(newState);
  }

  render() {
    /* Set the buttons list for the editor controls */
    const upperEditorButtonsList = [
      {
        id: 'add-new-item-button',
        onClick: this.showAddNewItemDialog,
        title: 'Add New Item',
        tooltip: 'Add a new item to your document.',
        type: 'button',
      },
    ];
    /* Set the CSS styling for the span element to hold the value of the item to be added */
    const hiddenSpanStyling = 'ajc-display-none';
    return (
      <React.Fragment>
        <HorizontalButtonList alignment="left" additionalUpperSpacing={this.props.additionalSpacingAboveControls} colour={this.props.colour}
          buttonList={upperEditorButtonsList} labelText="Click the button below to add new items to your document:" />      
        <span id="add-new-item-user-value" className={hiddenSpanStyling} />
  
        {/* Add new item dialog */}
        <AddNewItemDialog colour={this.props.colour} dialogContentAreaColour={this.props.dialogContentAreaColour} isDisplayed={this.state.dialogAddNewItem}
          handleClickClose={this.hideAddNewItemDialog} handleClickCancel={this.hideAddNewItemDialog} handleClickConfirmAddItem={this.handleClickConfirmAddItem} />
        
        {/* Loading new item dialog */}
        <LoadingNewItemDialog colour={this.props.colour} dialogContentAreaColour={this.props.dialogContentAreaColour} isDisplayed={this.state.dialogLoadingNewItem} />
      </React.Fragment>
    );
  }
}
ControlsEditorUpper.propTypes = {
  /** The functionality to be carried out to add a new item to the editor. */
  addEditorItem: PropTypes.func.isRequired,
  /** The colour / theme set to the document editor buttons and all relevant items in the preview page. */
  colour: PropTypes.oneOf([ 'grey', 'red' ]),
  /** The background colour for the content area of any dialogs rendered in the editor. By default this is set to white. */
  dialogContentAreaColour: PropTypes.oneOf([ 'grey', 'white', 'yellow' ]),
  /** The time allowance between a state change and a re-render occurring. */
  reRenderAllowance: PropTypes.number,
  /** Use additional spacing above the upper buttons. */
  additionalSpacingAboveControls: PropTypes.bool,
};
ControlsEditorUpper.defaultProps = {
  colour: 'grey',
  dialogContentAreaColour: 'white',
  reRenderAllowance: 100,
  additionalSpacingAboveControls: false,
};
export default ControlsEditorUpper;
