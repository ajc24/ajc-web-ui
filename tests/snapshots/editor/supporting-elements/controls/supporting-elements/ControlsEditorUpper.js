import 'jsdom-global/register';
import React from 'react';
import { TestDev } from 'ajc-jest-enzyme';
import ControlsEditorUpper from '../../../../../../src/ui-elements/editor/supporting-elements/controls/supporting-elements/ControlsEditorUpper';

describe('ControlsEditorUpper', () => {
  describe('Default props and rendering', () => {
    let jsonSnapshot;

    beforeAll(() => {
      jsonSnapshot = TestDev.createSnapshot(
        <React.Fragment>
          <ControlsEditorUpper addEditorItem={jest.fn()} />
        </React.Fragment>
      );
    });
    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });

  describe('Transferred props and rendering - All properties custom set', () => {
    let jsonSnapshot;

    beforeAll(() => {
      jsonSnapshot = TestDev.createSnapshot(
        <React.Fragment>
          <ControlsEditorUpper addEditorItem={jest.fn()} colour="red" dialogContentAreaColour="grey" additionalSpacingAboveControls={true} />
        </React.Fragment>
      );
    });
    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });

  describe('Dialog visible - Add new item dialog', () => {
    let jsonSnapshot;

    beforeAll(() => {
      const wrapper = TestDev.mount(
        <React.Fragment>
          <ControlsEditorUpper addEditorItem={jest.fn()} />
        </React.Fragment>
      );
      wrapper.find('ControlsEditorUpper').instance().showAddNewItemDialog();
      wrapper.update();
      const wrapperHtml = TestDev.html(wrapper);
      jsonSnapshot = TestDev.createSnapshot(wrapperHtml);
    });

    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });

  describe('Dialog visible - Loading new item dialog', () => {
    let jsonSnapshot;
    
    beforeAll(() => {
      const wrapper = TestDev.mount(
        <React.Fragment>
          <ControlsEditorUpper addEditorItem={jest.fn()} />
        </React.Fragment>
      );
      wrapper.find('ControlsEditorUpper').instance().showLoadingNewItemDialog();
      wrapper.update();
      const wrapperHtml = TestDev.html(wrapper);
      jsonSnapshot = TestDev.createSnapshot(wrapperHtml);
    });

    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });
});
