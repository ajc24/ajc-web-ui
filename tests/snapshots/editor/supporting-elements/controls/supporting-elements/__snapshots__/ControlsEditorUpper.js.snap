// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsEditorUpper Default props and rendering verifies the snapshot for the component 1`] = `
Array [
  <div
    className="ajc-flex-row ajc-spacing-bottom-small ajc-width-hd-inner ajc-text-black ajc-font-default ajc-flex-justify-left"
  >
    <h2
      className="ajc-font-paragraph-heading"
    >
      Click the button below to add new items to your document:
    </h2>
  </div>,
  <div
    className="ajc-flex-row ajc-spacing-bottom-small ajc-background-transparent ajc-width-hd-inner ajc-flex-justify-left ajc-spacing-top-reduced"
  >
    <button
      className="ajc-button ajc-font-default ajc-button-grey"
      id="add-new-item-button"
      onClick={[Function]}
      title="Add a new item to your document."
      type="button"
    >
      Add New Item
    </button>
  </div>,
  <input
    className="ajc-display-none"
    id="add-new-item-user-value"
    readOnly={true}
    type="text"
  />,
]
`;

exports[`ControlsEditorUpper Dialog visible - Add new item dialog verifies the snapshot for the component 1`] = `"<div class=\\"ajc-flex-row ajc-spacing-bottom-small ajc-width-hd-inner ajc-text-black ajc-font-default ajc-flex-justify-left\\"><h2 class=\\"ajc-font-paragraph-heading\\">Click the button below to add new items to your document:</h2></div><div class=\\"ajc-flex-row ajc-spacing-bottom-small ajc-background-transparent ajc-width-hd-inner ajc-flex-justify-left ajc-spacing-top-reduced\\"><button id=\\"add-new-item-button\\" class=\\"ajc-button ajc-font-default ajc-button-grey\\" type=\\"button\\" title=\\"Add a new item to your document.\\">Add New Item</button></div><input id=\\"add-new-item-user-value\\" type=\\"text\\" readonly=\\"\\" class=\\"ajc-display-none\\"><div id=\\"page-mask-add-new-item-dialog-id-root\\" class=\\"ajc-flex-row ajc-flex-justify-center ajc-page-mask ajc-width-4k\\"><div class=\\"ajc-flex-row ajc-flex-justify-center ajc-width-full-hd\\"><div class=\\"ajc-flex-row ajc-flex-justify-center ajc-width-hd\\"><div id=\\"page-mask-add-new-item-dialog-id-content\\" class=\\"ajc-flex-column ajc-flex-justify-start ajc-spacing-top-large ajc-spacing-bottom-large ajc-width-hd-with-padding ajc-text-white ajc-font-default\\"><div class=\\"ajc-flex-row ajc-flex-justify-center ajc-width-hd-inner\\"><div id=\\"add-new-item-dialog-id-root\\" class=\\"ajc-flex-column ajc-flex-justify-left ajc-width-dialog-parent\\"><div id=\\"add-new-item-dialog-id-title\\" class=\\"ajc-flex-row ajc-flex-justify-left ajc-dialog-title ajc-width-dialog ajc-text-white ajc-font-default ajc-background-grey-2\\"><h2 class=\\"ajc-font-dialog-title\\">Add New Document Item</h2></div><div id=\\"add-new-item-dialog-id-content-area-root\\" class=\\"ajc-flex-row ajc-flex-justify-left ajc-dialog-content ajc-width-dialog-content ajc-background-white\\"><div id=\\"add-new-item-dialog-id-content-area\\" class=\\"ajc-flex-column ajc-flex-justify-left ajc-text-black ajc-font-default\\"><div class=\\"ajc-flex-column ajc-flex-justify-left ajc-background-transparent ajc-spacing-bottom-small ajc-text-black ajc-font-default\\"><label for=\\"add-new-item-dropdown-menu-id\\" class=\\"ajc-row ajc-flex-justify-left ajc-spacing-bottom-small ajc-spacing-bottom-reduced ajc-label-standalone\\">Please select an item to be added to your document:</label><select id=\\"add-new-item-dropdown-menu-id\\" name=\\"add-new-item-dropdown-menu-id\\" class=\\"ajc-select ajc-font-default\\"><option value=\\"screenshot-with-caption\\" title=\\"Screenshot With Caption Text\\" selected=\\"\\">Screenshot With Caption Text</option></select></div></div></div><div id=\\"add-new-item-dialog-id-footer\\" class=\\"ajc-flex-row ajc-flex-justify-center ajc-dialog-footer ajc-width-dialog ajc-font-default ajc-background-grey-2\\"><button id=\\"confirm-add-new-item-button\\" class=\\"ajc-dialog-footer-button ajc-text-white ajc-menu-item-grey\\" type=\\"button\\">Confirm and Add</button><button id=\\"cancel-add-new-item-button\\" class=\\"ajc-dialog-footer-button ajc-text-white ajc-menu-item-grey\\" type=\\"button\\">Cancel and Exit</button></div></div></div></div></div></div></div><button id=\\"page-mask-add-new-item-dialog-id-close-button\\" class=\\"ajc-flex-row ajc-button ajc-button-page-mask-close ajc-text-white ajc-font-default ajc-button-grey\\" title=\\"Click to close\\">&nbsp;X&nbsp;</button>"`;

exports[`ControlsEditorUpper Dialog visible - Loading new item dialog verifies the snapshot for the component 1`] = `"<div class=\\"ajc-flex-row ajc-spacing-bottom-small ajc-width-hd-inner ajc-text-black ajc-font-default ajc-flex-justify-left\\"><h2 class=\\"ajc-font-paragraph-heading\\">Click the button below to add new items to your document:</h2></div><div class=\\"ajc-flex-row ajc-spacing-bottom-small ajc-background-transparent ajc-width-hd-inner ajc-flex-justify-left ajc-spacing-top-reduced\\"><button id=\\"add-new-item-button\\" class=\\"ajc-button ajc-font-default ajc-button-grey\\" type=\\"button\\" title=\\"Add a new item to your document.\\">Add New Item</button></div><input id=\\"add-new-item-user-value\\" type=\\"text\\" readonly=\\"\\" class=\\"ajc-display-none\\"><div id=\\"page-mask-loading-new-item-dialog-id-root\\" class=\\"ajc-flex-row ajc-flex-justify-center ajc-page-mask ajc-width-4k\\"><div class=\\"ajc-flex-row ajc-flex-justify-center ajc-width-full-hd\\"><div class=\\"ajc-flex-row ajc-flex-justify-center ajc-width-hd\\"><div id=\\"page-mask-loading-new-item-dialog-id-content\\" class=\\"ajc-flex-column ajc-flex-justify-start ajc-spacing-top-large ajc-spacing-bottom-large ajc-width-hd-with-padding ajc-text-white ajc-font-default\\"><div class=\\"ajc-flex-row ajc-flex-justify-center ajc-width-hd-inner\\"><div id=\\"loading-new-item-dialog-id-root\\" class=\\"ajc-flex-column ajc-flex-justify-left ajc-width-dialog-parent\\"><div id=\\"loading-new-item-dialog-id-title\\" class=\\"ajc-flex-row ajc-flex-justify-left ajc-dialog-title ajc-width-dialog ajc-text-white ajc-font-default ajc-background-grey-2\\"><h2 class=\\"ajc-font-dialog-title\\">Adding Your New Item</h2></div><div id=\\"loading-new-item-dialog-id-content-area-root\\" class=\\"ajc-flex-row ajc-flex-justify-left ajc-dialog-content ajc-width-dialog-content ajc-background-white\\"><div id=\\"loading-new-item-dialog-id-content-area\\" class=\\"ajc-flex-column ajc-flex-justify-left ajc-text-black ajc-font-default\\"><p class=\\"ajc-flex-row ajc-background-transparent ajc-page-content-text ajc-spacing-bottom-small ajc-text-black ajc-font-default ajc-flex-justify-left\\">Please wait just a moment. We are getting your new item ready for you.</p></div></div><div id=\\"loading-new-item-dialog-id-footer\\" class=\\"ajc-flex-row ajc-flex-justify-center ajc-dialog-footer ajc-width-dialog ajc-font-default ajc-background-grey-2\\"></div></div></div></div></div></div></div>"`;

exports[`ControlsEditorUpper Transferred props and rendering - All properties custom set verifies the snapshot for the component 1`] = `
Array [
  <div
    className="ajc-flex-row ajc-spacing-bottom-small ajc-width-hd-inner ajc-text-black ajc-font-default ajc-flex-justify-left ajc-spacing-top-large"
  >
    <h2
      className="ajc-font-paragraph-heading"
    >
      Click the button below to add new items to your document:
    </h2>
  </div>,
  <div
    className="ajc-flex-row ajc-spacing-bottom-small ajc-background-transparent ajc-width-hd-inner ajc-flex-justify-left ajc-spacing-top-reduced"
  >
    <button
      className="ajc-button ajc-font-default ajc-button-red"
      id="add-new-item-button"
      onClick={[Function]}
      title="Add a new item to your document."
      type="button"
    >
      Add New Item
    </button>
  </div>,
  <input
    className="ajc-display-none"
    id="add-new-item-user-value"
    readOnly={true}
    type="text"
  />,
]
`;
